### PR TITLE
API split of #2354

### PR DESCRIPTION
**Pull request by loevstroem**
Full API split from #2354

### CODE REVIEW

**Comment by pjedrzejewski** [comment]
Awesome, could you have a look at the failing build? Thank you for your work Kristian! :+1:

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php [hunk 1]
```diff
@@ -207,6 +213,55 @@ public function countByUserAndCoupon(UserInterface $user, CouponInterface $coupo
...
+                ->setParameter('createdAtFrom', $criteria['createdAtFrom'])
+            ;
+        }
+        if (!empty($criteria['createdAtTo'])) {
+            $queryBuilder
+                ->andWhere($queryBuilder->expr()->lte('o.createdAt', ':createdAtTo'))
+                ->setParameter('createdAtTo', $criteria['createdAtTo'])
+            ;
+        }
+        if (!empty($criteria['channel'])) {
```
Comment (stloyd) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php]: This seems to be unrelated.

File: src/Sylius/Bundle/CoreBundle/Form/Type/Api/OrderType.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ */
+class OrderType extends BaseOrderType
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('channel', 'sylius_channel_choice', array(
```
Comment (stloyd) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Api/OrderType.php]: Unrelated.

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php [hunk 3]
```diff
@@ -165,6 +165,12 @@ public function createFilterPaginator($criteria = array(), $sorting = array(), $
                 ->setParameter('paymentState', $criteria['paymentState'])
             ;
         }
+        if (!empty($criteria['channel'])) {
```
Comment (stloyd) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php]: Unrelated.

**Comment by michalmarcinkowski** [comment]
Great job! I needed this, thank you :+1:

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php (hunk 1 above)
Comment (loevstroem) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php]: Thanks for noticing Joseph

**Comment by saidul** [comment]
Promotion, Orders etc backend API should be protected before merge.

**Comment by loevstroem** [comment]
@pjedrzejewski I've updated code to fix tests but 1 build still fails https://travis-ci.org/Sylius/Sylius/jobs/48254064 Any idea why?

**Comment by pjedrzejewski** [comment]
I have restarted it, let's see now. @saidul @loevstroem Last thing before we merge this, remove the password and salt from user response. Also, we should define some default security. Maybe simply put the whole API behind ROLE_API? Later we can think about something smarter.

**Comment by pjedrzejewski** [comment]
Also, let's add a group with this ROLE_API role.

**Comment by loevstroem** [comment]
@pjedrzejewski The password and salt from user response is already excluded. ``/api`` now requires ROLE_API. Please review.

File: src/Sylius/Bundle/OrderBundle/Controller/OrderItemController.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        if (null === $orderId = $this->getRequest()->get('orderId')) {
+            throw new NotFoundHttpException('No order id given.');
+        }
+
+        if (!$order = $this->getOrderRepository()->find($orderId)) {
+            throw new NotFoundHttpException('Requested order does not exist.');
+        }
+
+        $orderItem = parent::createNew();
+        $orderItem->setOrder($order);
```
Comment (stloyd) [inline src/Sylius/Bundle/OrderBundle/Controller/OrderItemController.php]: Shouldn't this be done in listener?

File: src/Sylius/Component/Core/Model/OrderInterface.php [hunk 5]
```diff
@@ -55,6 +62,20 @@ public function getBillingAddress();
...
+     * Get the checkout state.
+     *
+     * @return string
+     */
+    public function getCheckoutState();
+
+    /**
+     * Set the checkout state.
+     *
+     * @param string $
```
Comment (stloyd) [inline src/Sylius/Component/Core/Model/OrderInterface.php]: Should be `$checkoutState`.

**Comment by pjedrzejewski** [comment]
Wooo, awesome, thanks for your effort Kristian! :+1:

**Comment by kayue** [comment]
@pjedrzejewski Will you use the new checkout controller to replace the flow bundle?

**Comment by pjedrzejewski** [comment]
@kayue No the api checkout controller is kinda temporary, I'd like to use Flow bundle for that as well, however api is stateless and flow depends on session atm. But I will eventually rework it to use statemachine as well. :)

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/main.yml [hunk 6]
```diff
@@ -90,7 +90,7 @@ fos_rest:
     format_listener:
         rules:
             - { path: '^/api', priorities: ['json', 'xml'], fallback_format: json, prefer_extension: true }
-            - { path: '^/', stop: true }
```
Comment (psyray) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/main.yml]: Hi, could you please take a look at this issue https://github.com/Sylius/Sylius/issues/2421 which was the consequence of the `stop: true` removal
